plugins/rich-editor/src/scripts/components/toolbars/ParagraphDropDown.tsx

```diff
@@ -0,0 +1,139 @@
+import Formatter from "@rich-editor/quill/Formatter";
+import ParagraphToolbarMenuItems from "@rich-editor/components/toolbars/pieces/ParagraphToolbarMenuItems";
+import CodeBlockBlot from "@rich-editor/quill/blots/blocks/CodeBlockBlot";
+import BlockquoteLineBlot from "@rich-editor/quill/blots/blocks/BlockquoteBlot";
+import SpoilerLineBlot from "@rich-editor/quill/blots/blocks/SpoilerBlot";
+import MenuItems from "@rich-editor/components/toolbars/pieces/MenuItems";
+import FocusWatcher from "@library/FocusWatcher";
+import DropDown from "@library/components/dropdown/DropDown";
+import { ButtonBaseClass } from "@library/components/forms/Button";
+import { heading2, heading3, blockquote, codeBlock, spoiler } from "@library/components/icons/editorIcons";
+import * as editorIcons from "@library/components/icons/editorIcons";
+
+interface IProps extends IWithEditorProps {
+    disabled?: boolean;
+    renderAbove?: boolean;
+    renderLeft?: boolean;
+}
+
+interface IState {
+    hasFocus: boolean;
+}
+
+/**
+ * Implemented ParagraphDropDown component, this is for mobile
+ */
+export class ParagraphDropDown extends React.PureComponent<IProps, IState> {
+    private quill: Quill;
+    private ID: string;
+    private componentID: string;
+    private menuID: string;
+    private buttonID: string;
+    private menuRef: React.RefObject<MenuItems> = React.createRef();
+    private formatter: Formatter;
+
+    constructor(props: IProps) {
+        super(props);
+
+        // Quill can directly on the class as it won't ever change in a single instance.
+        this.quill = props.quill;
+        this.formatter = new Formatter(this.quill);
+        this.ID = this.props.editorID + "paragraphMenu";
+        this.componentID = this.ID + "-component";
+        this.menuID = this.ID + "-menu";
+        this.buttonID = this.ID + "-button";
+        this.state = {
+            hasFocus: false,
+        };
+    }
+
+    public render() {
+        return (
+            <DropDown
+                id={this.ID}
+                className="mobileParagraphMenu-dropDown"
+                buttonContents={editorIcons.pilcrow()}
+                buttonBaseClass={ButtonBaseClass.ICON}
+                disabled={this.props.disabled}
+                renderAbove={this.props.renderAbove}
+                renderLeft={this.props.renderLeft}
+            >
+                <ParagraphToolbarMenuItems
+                    menuRef={this.menuRef}
+                    formatter={this.formatter}
+                    activeFormats={this.props.activeFormats}
+                    lastGoodSelection={this.props.instanceState.lastGoodSelection}
+                    onKeyDown={this.handlePilcrowKeyDown}
+                />
+            </DropDown>
+        );
+    }
+
+    /**
+     * Get the active format for the current line.
+     */
+    private get activeFormatIcon(): JSX.Element {
+        const { activeFormats } = this.props;
```

Comment:
This seem to be duplicated now across the 2 different versions of this. Could we:

- Make it it's own component. Maybe `ParagraphActiveIcon` and it just takes `activeFormats` as a prop. It __shouldn't__ need to be directly connect with the `withEditor` HOC.
- Replace this and the equivalent `activeFormatIcon` in the toolbar version of this.

Reply:
Good call